apollo-federation/src/merger/merge_input.rs

```diff
@@ -0,0 +1,605 @@
+                        code: HintCode::InconsistentInputObjectField.code().to_string(),
+                        message: format!(
+                            "Input object field \"{}\" will not be added to \"{}\" in the supergraph as it does not appear in all subgraphs: it is defined in {} but not in {}",
+                            dest_field.field_name,
+                            dest.type_name,
+                            present_subgraphs_str,
+                            missing_subgraphs_str
+                        ),
+                        locations,
+                    });
+                }
+                // Note that we remove the element after the hint/error because we access the parent in the hint message.
+                dest_field.remove(&mut self.merged)?;
+            }
+        }
+
+        // We could be left with an input type with no fields, and that's invalid in GraphQL
+        let final_input_object = dest.get(self.merged.schema())?;
+        if final_input_object.fields.is_empty() {
+            self.error_reporter.add_error(CompositionError::InternalError {
+            message: format!(
+                "None of the fields of input object type \"{}\" are consistently defined in all the subgraphs defining that type. As only fields common to all subgraphs are merged, this would result in an empty type.",
+                dest.type_name
+            ),
+        });
+        }
+
+        Ok(())
+    }
+
+    fn add_fields_shallow(
+        &mut self,
+        sources: &Sources<Node<InputObjectType>>,
+        dest: &InputObjectTypeDefinitionPosition,
+    ) -> IndexMap<
+        crate::schema::position::InputObjectFieldDefinitionPosition,
+        Sources<Component<InputValueDefinition>>,
+    > {
+        let mut added: IndexMap<
+            crate::schema::position::InputObjectFieldDefinitionPosition,
+            Sources<Component<InputValueDefinition>>,
+        > = IndexMap::default();
+        let mut fields_to_add: IndexMap<usize, HashSet<Option<Component<InputValueDefinition>>>> =
+            IndexMap::default();
+        let mut extra_sources: Sources<Component<InputValueDefinition>> = IndexMap::default();
+
+        // Process each subgraph
+        for (source_index, source) in sources.iter() {
+            let fields_set = fields_to_add.entry(*source_index).or_default();
+
+            if let Some(source_input) = source {
+                // Add all fields from this subgraph
+                for field in source_input.fields.values() {
+                    fields_set.insert(Some(field.clone()));
+                }
+            }
+
+            // Check if this subgraph defines the input type (even if source is None)
+            if let Some(subgraph) = self.subgraphs.get(*source_index) {
+                if subgraph.schema().get_type(dest.type_name.clone()).is_ok() {
+                    extra_sources.insert(*source_index, None);
+                }
+            } else {
```

Comment:
this seems like a change from JS behavior? (i.e. (in JS we only populate `extraSources` IF schema defines the type)[https://github.com/apollographql/federation/blob/main/composition-js/src/merging/merge.ts#L1290-L1298]))